### ISSUE TITLE
MutationalPatterns

### ISSUE THREAD

**Issue by roelj** [Bioconductor/Contributions#122]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/CuppenResearch/MutationalPatterns

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Subversion][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/source-control/
[6]: http://bioconductor.org/developers/how-to/git-mirrors/

**Comment by lawremi** [comment]
1. Contains yet another BED parser. Might be fun to compare all these different implementations.
2. Uses `for()` and `which()` too much.
3. `rename_chrom()` could be/use `seqlevelsStyle<-()`
4. Strangely uses `readRDS()` to read a file with an `.R` extension.
5. Should use formal list classes, e.g., `read_vcf()` could return a "GenomicRangesList". Btw, might consider a better name for that function; so similar to `readVcf()`.
6. Probably could just `import(ggplot2)`. Gets a bit noisy importing every symbol individually.
7. Consider using a SummarizedExperiment for storing mutation profiles, instead of just a matrix.
8. How does this compare to and/or complement the SomaticSignatures package?
9. For `plot_rainfall()` at least, instead of requiring the user to pass a BSgenome along with the VCF object, could use `genome(vcf)` to find and load the BSgenome automatically. Easier and less error-prone.
10. For example in `intersect_with_region()`, there must be better ways to check for compatibility of genomes than comparing just the seqlevel style.

Limiting myself to 10 points for each dose.

**Comment by roelj** [comment]
Hi @lawremi,

Thanks for your quick response. We are working hard to address all comments. At this point we have removed the `rename_chrom` in favor of `seqlevelsStyle`.  We removed the need to pass the reference genome in `plot_rainfall`, and we have updated and renamed `read_vcf` to `vcf_to_granges`. We also removed `bed_to_granges` in favor of `rtracklayer`'s `import` functionality.

We are running tests to make sure the next run will go fine. When we push the changes and increment the version number I will directly address each point again in a new comment.

Again, thanks for your effort. I expect to have a new build run ready some time tomorrow.

**Comment by lawremi** [comment]
Are you aware of the VRanges class in VariantAnnotation? It gives you a GRanges form of a VCF, with formal representation of the alleles, read counts, etc. Also, `read_vcf()` does not seem to handle multi-sample VCF files.  You could just expect the user to call `readVcfAsVRanges()` each file in a loop over the files, potentially stacking them into one object (VRanges has a formal `sampleNames` field to keep things straight).

Like:
```
vr <- stackSamples(List(lapply(files, readVcfAsVRanges)))
```

**Comment by roelj** [comment]
I wasn't aware of it.  Thanks for pointing me to it. Yet another thing to look into.

I suppose that with the "Uses `for()` and `which()` too much." comment, you mean it would be better to adopt a functional programming style over a procedural style (use higher-order functions instead?)?

**Comment by lawremi** [comment]
Basically, yes. For `which()`, it's usually good enough to use logical subscripting, no need to convert to integer subscripts.

**Comment by roelj** [comment]
Thanks for your responses so far. Here's what we've done so far (I left out a few points we will address a bit later):

> 1.  Contains yet another BED parser. Might be fun to compare all these different implementations.

We removed `bed_to_granges` and use `rtracklayer`'s `import` in the examples and the vignette.

> 2. Uses `for()` and `which()` too much.

I'd like to postpone these changes to a later time, because what we now have has been tested by users. Possibly in combination with adding unit tests, we can transition away from the procedural style.

> 3.  `rename_chrom()` could be/use `seqlevelsStyle<-()`

We removed `rename_chrom()` because `seqlevelsStyle` does what we need to do here.  We adjusted the examples and the vignette accordingly.

> 4.  Strangely uses readRDS() to read a file with an .R extension.

It reads serialized data for making the examples easier and quicker to follow. I changed the `.R` extension to `.rds`.  Now it shouldn't be so strange anymore :)

> 5.  Should use formal list classes, e.g., read_vcf() could return a "GenomicRangesList". Btw, might consider a better name for that function; so similar to readVcf().

We renamed `read_vcf()` to `vcf_to_granges()` and that function returns a `GRangesList` instead of a regular `list`. In addition, this function applies a naming style that can be passed with the newly introduced `style` parameter. The default is "UCSC" as we used throughout our examples and in the vignette.

> 6.  Probably could just `import(ggplot2)`. Gets a bit noisy importing every symbol individually.

Changed accordingly.

> 9.  For `plot_rainfall()` at least, instead of requiring the user to pass a BSgenome along with the VCF object, could use `genome(vcf)` to find and load the BSgenome automatically. Easier and less error-prone.

Thanks for this useful suggestion! We removed the `ref_genome` parameter.


In addition to addressing these points, we want to further change the vignette to be more clear.

**Comment by roelj** [comment]
On Windows, there seem to be a failures outside of my ability to solve:

```
Error in loadNamespace(i, c(lib.loc, .libPaths()), versionCheck = vI[[i]]) : 
  there is no package called 'digest'
Error: package 'pkgmaker' could not be loaded
Execution halted
```

Is there anything I can do about this, or can I ignore this warning?

**Comment by lawremi** [comment]
1. How about `readVcfsAsGRanges()` instead of `vcf_to_granges()`?
2. `get_muts()` assumes that there is only one alt per row. That is only guaranteed to be true for _ExpandedVcf_, not _CollapsedVcf_. I wonder though if this function should even exist. It's almost easier and clearer to write `paste0(ref(vcf), ">", alt(vcf))`.
3. `get_types()` is too generic a name. Typically in R, we omit the `get_` from the getters. Maybe something like `mutation_types()`?
4. `get_mut_context()` is just `getSeq(vcf+1, genome)`, although see comment below.
5. `mut_type_occurrences()` could be considerably simplified by using factors (to avoid missing columns) and by combining the input data into one object and computing a two-way table. A previous recommendation was to simply combine the data across samples, using a _VRanges_ and its "sampleNames" column.
6. To reiterate, this package seems very similar to SomaticSignatures. Please consider at least citing that work in your vignette. It would be helpful to explain to users the difference and how the two might complement each other. I invite @juliangehring to comment.

General comments to those listening:
1. Looking at the vignette, maybe GenomeInfoDb needs a function `keepAutosomes()` to go along with `keepStandardChromosomes()`.
2. Could `getSeq()` be made smart enough to automatically find the BSgenome package associated with `genome(x)`, so that it's not necessary to pass the genome?

Probably more later.

**Comment by FrancisBlokzijl** [comment]
Hi @lawremi,

I'd like to respond to your question:
How does this compare to and/or complement the SomaticSignatures package?

MutationalPatterns offers many more functionalities than SomaticSignatures

- SomaticSignatures provides functions for genomic context determination and de novo signature extraction using NMF decomposition of 96 trinucleotide count matrices.
- MutationalPatterns offers these functionalities as well, yet plotting is a bit different. Importantly, MutationalPatterns offers the functionality to extract signatures de novo from a 192 feature matrix, with 96 trinucleotides X 2 transcriptional strands. This allows assessment of transcriptional strand bias of mutational signatures, which is very important to characterize the underlying mutational mechanism.

More features of MutationalPatterns that are not in SomaticSignatures:
- Function to determine the optimal contribution of previously identified signatures, e.g. cosmic cancer signatures, to reconstruct the mutational profile of just a single sample. This is useful for two reasons: (1) for NMF you need many samples with distinct mutational profiles, as it relies on dimensionality reduction. (2) In order to further characterize “known” mutational signatures and find the underlying mutational mechanisms, this function can be used to determine the contribution of these signatures in different samples, e.g. normal cells, or cells with defective DNA repair mechanisms etc.
- Plotting function and statistical test for enrichment or depletion of mutations in any (publicly available) annotated genomic region, such as transcription factor binding site or “open chromatin”, this is useful for characterization of mutational mechanisms
- Statistical test and plotting function for transcriptional strand bias in mutation catalogs
- Plotting function to visualize the difference between two 96 mutational profiles and calculate RSS

Hope this answers your question!

Kind regards,
Francis Blokzijl

**Comment by FrancisBlokzijl** [comment]
**1. How about readVcfsAsGRanges() instead of vcf_to_granges()?

2. get_muts() assumes that there is only one alt per row. That is only guaranteed to be true for ExpandedVcf, not CollapsedVcf. I wonder though if this function should even exist. It's almost easier and clearer to write paste0(ref(vcf), ">", alt(vcf)).**

In the function read_vcf, which was changed into vcf_to_granges, does a little more than that just parsing the vcf. The function also checks and excludes postions with (1) indels and (2) multiple alternative alleles. So for your point 
1., I'm okay with changing it to readVcfsAsGRanges(), but do you think this is illustrative for what the function does? 
2. So the multiple alternative alleles are not in there anymore, so that's no problem. I agree that the function get_muts is not very fancy as it can be done in 1 line of code. I liked having it as a function and I use it in get_type_context for example. If you prefer, we can not export the function.

**get_types() is too generic a name. Typically in R, we omit the get_ from the getters. Maybe something like mutation_types()?**

We will rename the functions according to your suggestions.

**Comment by roelj** [comment]
@lawremi Would you prefer us to change all function names to camelCasing, as your suggestion for `readVcfsAsGRanges()`, in addition to getting rid of `get_` for getters?

**Comment by lawremi** [comment]
Thanks for explaining the improvements over SomaticSignatures. It might be to the user's benefit to include that explanation in the vignette somewhere. When a user is looking for a package to meet a need, it can be difficult to distinguish packages. 

I'm not sure why you are excluding positions with multiple alts. Since you're aggregating to the trinucleotide level, is it such a big deal to include those as separate records, i.e., to just `expand()`? 

It is tough to summarize a high-level function in a single name. Maybe `read_snvs()`? It's still not clear why it's not using _VRanges_, since it's basically designed for this problem.

`get_muts()` man page says it is expecting a CollapsedVcf, but I think you're actually passing a GRanges derivative of that. Again, it would be nice to have this be more formal, and the _VRanges_ data structure is already there for you. As _VRanges_ is expanded, there is no need to deal with the list of alts. Each entry point in the API can be called independently of the others, and integration arises through contract, i.e., formally defined data structures. `get_muts()` should not assume the user has constructed a _GRanges_ with `vcf_to_ranges()`.

I don't care about the casing.

**Comment by roelj** [comment]
Hi @lawremi 

> How about readVcfsAsGRanges() instead of vcf_to_granges()?

I renamed it to `read_vcfs_as_granges`, keeping the naming style for our package consistent. 

> get_types() is too generic a name. Typically in R, we omit the get_ from the getters. Maybe something like mutation_types()?

I applied your proposed change. Thanks for the suggested naming. I like it.
I can/will apply similar changes to the other `get_` functions.

> get_mut_context() is just getSeq(vcf+1, genome), although see comment below.

I have to perform more tests to determine the impact of using `getSeq` instead.

> mut_type_occurrences() could be considerably simplified by using factors (to avoid missing columns) and by combining the input data into one object and computing a two-way table. A previous recommendation was to simply combine the data across samples, using a VRanges and its "sampleNames" column.

Thanks for the pointer to factors.  I changed it to a one-liner :) (well.. almost..)

**Comment by roelj** [comment]
Again, the warnings on the Windows build are beyond my control:

```
Error in loadNamespace(i, c(lib.loc, .libPaths()), versionCheck = vI[[i]]) : 
  there is no package called 'digest'
Error: package 'pkgmaker' could not be loaded
Execution halted
```

**Comment by juliangehring** [comment]
> We have deposited our manuscript in bioRxiv. http://biorxiv.org/content/early/2016/08/30/071761 
In the manuscript we do refer to the SomaticSignatures work of @juliangehring. We will also include this reference in the package description.

The preprint list the paper in the reference list, but does not refer to it in the text itself. As such, it does not help much in understanding the similarities and differences between the two for the reader/user.

**Comment by juliangehring** [comment]
I agree with @lawremi that explaining the overlap with existing packages and the new features that mutationalPatterns adds would be very helpful for a user. Since there are already some R packages with similar scopes and methodologies, I would suggest that this discussion should include SomaticSignatures, deconstructSigs and pmsignature (not meant to be a comprehensive list though).

**Comment by roelj** [comment]
In the latest version I added the explanation of the differences between `SomaticSignatures` and `MutationalPatterns`.  I don't know about the other packages (`deconstructSigs` and `pmsignature`), so I cannot usefully compare them.

I also renamed the remaining `get_` functions to appropriate alternatives (suggestions are still welcome).

I assume the Windows build will signal the same warnings again.  I posted about this on the bioc-devel mailing list to hopefully get this sorted out quickly.

Even though we still have some code refactoring to do, I hope we can move on so that we can make it in the 3.4 release.  The function names (the API) should now be stable (unless there are suggestions for the naming in the next comments), so I believe it's safe to move on.

Thanks again for your time.

**Comment by vobencha** [comment]
Some general package comments:
- Packages in 'Depends' should be fully imported (e.g., import(graphics)) in NAMESPACE. If you want to selectively import the package should be in 'Imports'. Please fix this for graphics, stats, NMF.
- Remove manual_code.R or move it to inst/scripts/.
- Remove MutationalPatterns.Rproj.

Valerie

**Comment by vobencha** [comment]
Checking in. Are you planning to submit a new version? Deadline for inclusion in the next release is October 12.
Valerie

**Comment by roelj** [comment]
Yes @vobencha, I am planning to submit a new version ASAP.  I am working on a bugfix as well which is delaying the process a bit.

**Comment by roelj** [comment]
I removed the RStudio project file and `manual_code.R` as requested, and I fixed the Imports in `NAMESPACE`.

**Comment by roelj** [comment]
For our current user's convenience, I added error messages pointing to the "new way of doing things".  For example, the `bed_to_granges` function has been removed earlier, and three lines replace the function (using `rtracklayer`'s `import` function and unifying the `seqlevelsStyle`s).

I believe we now have a package that integrates well with existing Bioconductor packages, that has sensible function names, and "just works".  So I think it should be ready for inclusion into Bioconductor. :)
